### ISSUE TITLE
How to skip DynamoDB Table creation if already exists

### ISSUE THREAD

**Issue by rampatina** [aws/aws-cdk#6762]
### The Question
How can we skip if the Dynamo Table creation if already existing in the stack, otherwise always create a new table.
or Is there any way to catch the error and proceed with the next steps? (Similar to Try/Catch)

### Environment

  - **CDK CLI Version:** 1.27.0
  - **Module Version:** 1.20.0
  - **OS:** Macintosh
  - **Language:** TypeScript

**Comment by RomainMuller** [comment]
We have a `SingletonFunction` construct that does something similar for lambda `Functions`. You can find the source of it at [@aws-cdk/aws-lambda/lib/singleton-lambda.ts](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/singleton-lambda.ts).

Conceptually, it creates the resource in the root `Stack` (instead of in some construct's scope). A UUID is assigned to reduce the risk of collisions with a legitimate construct scoped to the `Stack` as well... Then we use `tryFindChild` to tell if the singleton already exists, creating it otherwise.

You should be able to easily implement the same pattern for a DynamoDB Table.

**Comment by rampatina** [comment]
Thanks for the reply. Do we have any way to check the existence of the DynamoDB table in an AWS environment through CDK?

**Comment by Douglas-Scott** [comment]
I believe this is possible with a CustomResource.

https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-cloudformation.CustomResource.html

**Comment by mattiLeBlanc** [comment]
@Douglas-Scott Sorry, but that documentation page is far from helpful.
Is there any example available how to check  if the table does exist? We use `RemovalPolicy.RETAIN` on the DB so when we remove our stack but not our DB, it would be nice if it could skip DB creation and just continue.

**Comment by mattiLeBlanc** [comment]
I create on solution which you might find hacky, but it works fine in our CDK stack deploy script:
```
import * as Dynamodb from 'aws-sdk/clients/dynamodb';
import { SharedIniFileCredentials } from 'aws-sdk';
   ...

  async checkIfDynamoDbExists(name: string ) {
    const credentials = new SharedIniFileCredentials({profile: this.user});
    const dynamodb = new Dynamodb({
      region: this.config.region,
      credentials: this.user ? credentials : undefined
    });
    const tables = await dynamodb.listTables().promise();
    return tables!.TableNames!.indexOf(name) > 0;
  }
```
It will return me a true or false, and based on that I either create a table or get it:

```
 const tableExists = await this.resourceService.checkIfDynamoDbExists(applicationName);
    let dynamoDBTable: Table;
    if (!tableExists) {
      dynamoDBTable = new DynamoDbConstruct(this, `${id}-dynamoDb`, {
        applicationName
      }).table;
    } else {
      dynamoDBTable = Table.fromTableName(this, 'Table', applicationName) as Table;
    }
```

**Comment by malpaso** [comment]
Why is this issue closed? Seems like a pretty big missing piece for IaC. How is a CDK stack that creates an API, crud Lambdas and a DynamoDB supposed to re-deploy?

**Comment by mickog** [comment]
I agree this is crazy that this issue is closed.

**Comment by mattiLeBlanc** [comment]
So a downside of that workaround I proposed is that if you recreate your stack, and the table needs to be assimilated, it will not be part of the stack by just using the `fromTableName` solution.
You actually have to import the CFN resource of your table back into your new stack using an import in the CF dashboard.

If you your stack has created the table in the first place, it will not throw an error on redeploy. It knows that the table exists. Only if you have recreated your stack, the table would be an orphan. And you have to bring it into your family.

**Comment by mickog** [comment]
I've created another issue here https://github.com/aws/aws-cdk/issues/11208 raised it as a bug cause it sounds like it is to me but looks like the stance on it is that when a stack is pulled down it only retains resources to prevent data loss and is not meant to be used again. I thought it was common enough to pull down stacks though from time to time for a fresh deployment it can be great to fix something when having problems trouble shooting an issue. The old turn it off and back on trick

**Comment by tempestrock** [comment]
>     return tables!.TableNames!.indexOf(name) > 0;

This may need to be `...indexOf(name) >= 0` as the index of the table name is 0 if it exists.

**Comment by chiller-whale** [comment]
@RomainMuller this suggestions seems like it would prevent the same resource being created in the same stack, but the issue is happening between deployments of the same stack.

For example:
1. Create a stack that creates a table named `TableA`
2. Deploy that stack
3. Deploy the stack again

Expected: The stack is deployed successfully when deploying on step 3.
Observed: When deploying on step 3, an error is thrown because `TableA` already exists.

Should this use case be a separate issue?

**Comment by tmerse** [comment]
> dynamoDBTable = Table.fromTableName(this, 'Table', applicationName) as Table;

It is not even safe to assume this works, as list

> const tables = await dynamodb.listTables().promise();
>     return tables!.TableNames!.indexOf(name) > 0;

It is not even safe to assume this will always work, as the result of listTables() is paginated.

**Comment by bigboiblue** [comment]
I've created a solution which might seem a bit long, but works for any construct as far as I'm aware (and will pay off if you do this to multiple resources).

First of all I've assumed this is being done because the removalPolicy is retain, though if this is not the case, naming schemes can easily be modified.

Unfortunately this solution involves tags. If this doesn't float your boat, sorry.

The essence is that you search for aws resources with a predefined tag key. We use the ARN from the aws resource to get the id of the resource. If the id matches one of the ids of the cdk constructs, skip the creation of that construct:

(I cba to extract stuff out so here is the entire file)

```
import { Construct } from "constructs";
import { RemovalPolicy, Stack, StackProps, Tags} from "aws-cdk-lib"
import * as ecr from "aws-cdk-lib/aws-ecr"
import { config as AWSConfig, ResourceGroupsTaggingAPI } from "aws-sdk"
AWSConfig.update({region:'eu-west-2'});

const RETAIN_TAG = "retain-removal-policy"

export default class BaseStack extends Stack {
    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props)

        this.initStatefulConstructs()
    }

    /**
     * Stateful constructs will generally be retained (cdk destroy will keep the constructs in AWS)
     */
    async initStatefulConstructs() {
        const resources = [
            new ecr.Repository(this, "api-repo", {
                repositoryName: "api-repo",
                imageTagMutability: ecr.TagMutability.MUTABLE,
                removalPolicy: RemovalPolicy.RETAIN
            })
        ]
       
        let deployedRetainedIds: string[] = await this.getDeployedRetainedIds()

        resources.forEach(x => {
            Tags.of(x).add(RETAIN_TAG, "")

            const id = x.node.id
            if (deployedRetainedIds.includes(id)){
                console.log(`${id} already exists - skipping creation`)
                this.node.tryRemoveChild(id)
            }
        }) 
    }

    getIdFromARN(arn: string): string {
        let name = arn.substring(arn.search(/:.*$/) + 1)
        name = name.substring(name.search(/\/.*$/) + 1)
        return name
    }

    async getDeployedRetainedIds() {
        let deployedRetainedIds: string[] = []
        const api = new ResourceGroupsTaggingAPI()

        let nextPage = ""
        do{
            const ret = await api.getResources({
                TagFilters: [{ Key: RETAIN_TAG }],
                PaginationToken: nextPage
            }).promise()
            nextPage = ret.PaginationToken ?? ""
            const ids = ret.ResourceTagMappingList?.map(x => this.getIdFromARN(x.ResourceARN ?? '')) ?? []
            deployedRetainedIds = [...deployedRetainedIds, ...ids]
        } while(nextPage !== "")
        return deployedRetainedIds
    }
    
}
```

Although the async function is not awaited, it will be executed before program exit.

Just add any constructs you want to skip creation of (when existing) to the resources array.

Hope someone finds this helpful

**Comment by kevin-rlty** [comment]
Hello, any update on that issue?
Currently I'm locally storing tables per env that don't need update into a json file (far to be optimal)
Can't we update an existing table? Like adding scalability parameters or stream options, .. .?

**Comment by balchanj** [comment]
Why is this closed? It seems weird that most other constructs can be changed and are ignored if there isnt any change in a stack, but for some reason dynamodb tables cause it to fail. I dont see how hacky workarounds are a viable answer to a bad feature.

**Comment by wz2b** [comment]
Is this the same strategy I would use for cognito user pools?  I have the same problem.  If there's a user pool already there by this name I don't want to create another one.  This way I can turn off delete protection and have everything still work.

**Comment by gabrielcho** [comment]
Bump, we need an update on this, it's been almost 3 years since this issue was generated.

**Comment by deepanprabhu** [comment]
+1

**Comment by lancedolan** [comment]
This appears to be a feature request, that is valuable to the community, and is in fact behavior many assume it would have already. 

The ask is simple, however difficult the implementation may be: when deploying a stack, have the option to adopt persistent resources (DDB, S3, etc) during name collision rather than erroring out. 

Retaining persistent resources during stack deletion is definitely better than deleting them, so kudos to AWS CFN/CDK teams for that, but if we cannot re-adopt them then the dev is left in an encumbered position and re-deploying the stack around those resources is a very likely solution they want. 

Also, if can't control whether they are retained during stack deletion, and cannot re-adopt the resources on re-deploy, then tearing down a stack and deploying in "dev" account (often necessary during testing) becomes tedious and expensive.

**Comment by aiden-sobey** [comment]
This issue appears to be happening in CDK v2, albeit inconsistently. Does anyone know what triggers this?

**Comment by wascou** [comment]
Another brick in the wall.

I created a stack with CDK, re-arranged the stack so the code with the DynamoDB Tables are in a construct now, pushing the stack: 

> TableXXX already exists in stack arn:aws:cloudformation:xxx:xxx:stack/xxx

What is the work around ? Back up and restore?

**Comment by wz2b** [comment]
> What is the work around ? Back up and restore?

My personal workaround was to do it manually, which causes a stack drift that I then repaired.  It kind of sucked.

**Comment by mattiLeBlanc** [comment]
You can import orphan resources via Cloudformation back into your stack.
To do so, export the CF config to yaml.

Then generate your local CF config with “cdk synth” and extract the part
that references the table and copy that into the export from online.
Than you can import that yaml file and reassign the resources into the stack

**Comment by FelixRelli** [comment]
This should not be closed. The workaround is not a viable option as the table is not under CF management after import. Doing it by hand via the UI is a huge pain.

**Comment by aalobai1** [comment]
I agree please fix this 🙏

**Comment by dmarkowvw** [comment]
+1

**Comment by rlees85** [comment]
I think for me its just the defaults were massively confusing. I just wanted a DynamoDB table but if I deployed the stack more than once it just failed as it can't create a table that already exists, despite every other object in the stack (SQS,SNS,Lambda,etc) all being idempotent.

It looks like DynamoDB it set to retain by default, that means CloudFront just abandons its state as soon as its created.

I just had to add this line to "fix" it.
```
  const myTable = new dynamodb.Table(
    ins,
    tableKeyId,
    {
      ...
      removalPolicy: RemovalPolicy.RETAIN_ON_UPDATE_OR_DELETE // <- This line
    }
  );
```

There is still going to be a lot of manual intervention if the stack/table ever needs to be deleted and re-created but this is at least a much better position than before

**Comment by orfisko** [comment]
This is terrible. If the dynamodb table name is there it should simply include it in the stack.... Another serverless nightmare

**Comment by not-an-llm** [comment]
This is how to fix and get everything back to normal.

1. `cdk synth` to generate the template.
2. Copy the template `cdk.out/<stack-name>.template.json`.
3. Edit the template to remove unrelated resources; keep only the "Resource" items that are orphaned.
4. Go to console > CloudFormation > Stacks. If the stack is in rollback state, delete it.
5. Click on "Create Stack" > "With existing resources (import resources)"
6. When prompted, upload the template file.
7. Fill in the resource names and stack name. Make sure you get it right.
8. If you see "Cannot execute a change set that is pending creation", then just wait a second and click submit again

This will execute a change set to import the resources into the stack. Once the import is complete, you should be good to redeploy the stack. No need to change your CDK code.

This is pretty awful devex IMO though.
